pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -103,21 +98,36 @@ func (r *OperatorStepResolver) ResolveSteps(namespace string, _ SourceQuerier) (
 	updatedSubs := []*v1alpha1.Subscription{}
 	bundleLookups := []v1alpha1.BundleLookup{}
 	for name, op := range operators {
-		// TODO: added "is default channel" to sourceinfo out
-		// of convenience, which breaks map key equality here
-		// because it requires information that can't be
-		// gleaned from subscriptions alone. need to revisit
-		// TODO: this check also doesn't properly account for
-		// subscriptions made without a catalog specified,
-		// which means the sourceinfo won't match for the
-		// realized operator
+		// Find an existing subscription that resolves to this operator.
+		var (
+			existingSubscription *v1alpha1.Subscription
+			alreadyExists        bool
+		)
 		sourceInfo := *op.SourceInfo()
-		sourceInfo.DefaultChannel = false
-		_, isAdded := add[sourceInfo]
-		existingSubscription, subExists := subMap[sourceInfo]
+		for _, sub := range subs {
+			if sub.Spec.Package != sourceInfo.Package {
+				continue
+			}
+			if sub.Spec.Channel != "" && sub.Spec.Channel != sourceInfo.Channel {
+				continue
+			}
+			subCatalogKey := registry.CatalogKey{
+				Name:      sub.Spec.CatalogSource,
+				Namespace: sub.Spec.CatalogSourceNamespace,
+			}
+			if !subCatalogKey.Empty() && !subCatalogKey.Equal(sourceInfo.Catalog) {
+				continue
+			}
+			alreadyExists, err = r.hasExistingCurrentCSV(sub)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("unable to determine whether subscription has a preexisting CSV")
```

Comment:
nit: this isn't relevant to your PR, but it would be nice if we could create another abstraction for the results of this function so we don't return 4 different things, which looks a bit funny. Something like a `ResolutionResult` which encapsulates all the existing types. Just a thought.

Reply:
I agree with you, I would like to see the output eventually be a set of manifests to apply -- so the contract would basically be current state in, resolved state out.